## 11. Some real maths on Series

In [2]:
import pandas as pd

google = pd.read_csv('csv/google.csv', squeeze=True)

google.head()

0    50.12
1    54.10
2    54.65
3    52.38
4    52.95
Name: Stock Price, dtype: float64

### Counting total number of valid rows (w/o NaN)

In [4]:
google.count()

3012

Python's builtin len function also works similarly. Only main difference is **count()** counts only the valid **(Non NaN)** rows while len counts total number of rows in the series.

In [6]:
len(google) # There are no null values in this Series.

3012

### Calculating sum, prod, mean, median, mode, std (standard deviation), min, max

In [7]:
google.sum()

1006942.0

In [8]:
google.prod() # inf means Infinite. Too big value to calculate.

inf

In [10]:
google.mean() # Average of all values

334.3100929614874

In [11]:
google.median() # Median of the series

283.315

In [12]:
google.mode() # Most often ocurred value

0    291.21
dtype: float64

In [13]:
google.std() # Standard deviation of the series

173.18720477113115

In [14]:
google.min() # Minimum value of the series

49.95

In [15]:
google.max() # Maximum value of the series

782.22

### Finding the index of the maximum and minimum value
This can be done either by sorting or using **idxmax** and **idxmin** methods.

First we will see the sorting method that we learnt in earlier lessons.

In [16]:
google.head()

0    50.12
1    54.10
2    54.65
3    52.38
4    52.95
Name: Stock Price, dtype: float64

In [19]:
google.sort_values(ascending=False).head(1) # Sorting in descending order to find out index of maximum value

3011    782.22
Name: Stock Price, dtype: float64

In [20]:
google.sort_values().head(1)

11    49.95
Name: Stock Price, dtype: float64

So, **11** is the index of the minimum value and **3011** is the index of the maximum value. Lets find out the same thing using **idxmax** and **idxmin** methods.

In [21]:
google.idxmax()

3011

In [22]:
google.idxmin()

11

We can use the above logic in reverse fashion as well. Lets say we need to find the value of the min/max value in the series without using **.max()** and **.min()** methods.

In [23]:
google[google.idxmin()] # Similar to google.min()

49.95

In [24]:
google[google.idxmax()] # Similar to google.max()

782.22

### Calculating the number of unique values in the series
This is similar to the builtin **.count()** method (non-technically). But since we are going to count the repetitions in the values we need to use **value_counts** method.

In [28]:
pokemon = pd.read_csv('csv/pokemon.csv', squeeze=True, index_col='Pokemon')

pokemon.head(5)

Pokemon
Bulbasaur     Grass
Ivysaur       Grass
Venusaur      Grass
Charmander     Fire
Charmeleon     Fire
Name: Type, dtype: object

Lets say we need to find out or count how many Pokemons are of 'Psychic' type, we can use this method.

In [29]:
pokemon.value_counts() # Sorting in descending order, with max count at top.

Water       105
Normal       93
Grass        66
Bug          63
Psychic      47
Fire         47
Rock         41
Electric     36
Ground       30
Dark         28
Poison       28
Fighting     25
Dragon       24
Ghost        23
Ice          23
Steel        22
Fairy        17
Flying        3
Name: Type, dtype: int64

In [31]:
pokemon.value_counts(ascending=True) # Sorting in ascending order.

Flying        3
Fairy        17
Steel        22
Ice          23
Ghost        23
Dragon       24
Fighting     25
Poison       28
Dark         28
Ground       30
Electric     36
Rock         41
Fire         47
Psychic      47
Bug          63
Grass        66
Normal       93
Water       105
Name: Type, dtype: int64

## 12. Apply method
This methods applies a function on individual element of the Series. So lets define a function that we want to apply to our **pokemon** series.

In [37]:
def pokemon_function(value: str) -> None:
    """Perform .apply() method demo"""
    if value == 'Grass':
        return 'Ghass-puss'
    elif value == 'Fire':
        return 'Garam'
    elif value == 'Water':
        return 'Paani'
    elif value == 'Rock':
        return 'Dagad'
    elif value == 'Ground':
        return 'Maathi'
    elif value == 'Electric':
        return 'Jhatka'
    else:
        return 'Aaltu-Faaltu'
    
    
pokemon.apply(pokemon_function)

Pokemon
Bulbasaur     Ghass-puss
Ivysaur       Ghass-puss
Venusaur      Ghass-puss
Charmander         Garam
Charmeleon         Garam
                 ...    
Yveltal       Altu-Faltu
Zygarde       Altu-Faltu
Diancie            Dagad
Hoopa         Altu-Faltu
Volcanion          Garam
Name: Type, Length: 721, dtype: object

## 13. Map method
We use **.map()** method when we have 2 different Series and we need to map them together based on some common factors, similar to **VLOOKUP** in MS Excel.

In [11]:
pokemon_types = pd.read_csv('csv/pokemon.csv', squeeze=True, index_col='Pokemon')

pokemon_types.head()

Pokemon
Bulbasaur     Grass
Ivysaur       Grass
Venusaur      Grass
Charmander     Fire
Charmeleon     Fire
Name: Type, dtype: object

In [12]:
pokemon_names = pd.read_csv('csv/pokemon.csv', squeeze=True, usecols=['Pokemon'])

pokemon_names.head()

0     Bulbasaur
1       Ivysaur
2      Venusaur
3    Charmander
4    Charmeleon
Name: Pokemon, dtype: object

In [13]:
pokemon_names.map(pokemon_types)

0        Grass
1        Grass
2        Grass
3         Fire
4         Fire
        ...   
716       Dark
717     Dragon
718       Rock
719    Psychic
720       Fire
Name: Pokemon, Length: 721, dtype: object

Here, we used the pokemon names from the **pokemon_names** series and mapped to index of the **pokemon_types** series. 